README.md
# Switch Button - Toggle

<div align="center">
<img src="https://i.imgur.com/2Vz7bni.gif" alt="Preview do projeto" width="40%"/>
</div>

```
Com TypeScript o/
```

### <p align="center"> 🧪 Veja a <a href="https://switch-button-phi.vercel.app/" target="_blank">demonstração aqui</a></p>

## 🚀 Começando

- É só dar um fork no repo e usar ♡

### 📋 Pré-requisitos

- Instalar styled-components

```
yarn add styled-components
```

### 🔧 Como implementar

Parâmetros

```ts
  // Não obrigatório
  disabled?: boolean; // default false
  variant: 'vertical' | 'horizontal'; // default horizontal
  isActive: boolean; // default false
  onClick?: () => void;
```

```jsx
<SwitchButton
  variant="vertical"
  isActive={valueState}
  disabled
  onClick={() => stateHook((old) => !old)}
/>
```

### 🔧 Alterar cores

- Acessar arquivo styles.ts
  - linha 17 e 18 é responsável pela cor do fundo do botão inativo
  - linha 43 e 44 é responsável pela cor do círculo inativo
  - linha 56 e 57 é responsável pela cor do círculo ativo
  - linha 59 e 60 é responsável pela cor do fundo do botão ativo

## ✒️ Suporte

- **Wesley Souza** - _Full Stack Dev_ - [iwsouza](https://github.com/iwsouza)


CHANGELOG.md
## [1.1.2](https://github.com/renatonerijr/switch-button-deploy/compare/v1.1.1...v1.1.2) (2023-01-28)


### Bug Fixes

* **fix:** Change main ([7cf03e1](https://github.com/renatonerijr/switch-button-deploy/commit/7cf03e19ac4d798377256ba5e73acc11ced51a8d))

## [1.1.1](https://github.com/renatonerijr/switch-button-deploy/compare/v1.1.0...v1.1.1) (2023-01-28)


### Bug Fixes

* **fix:** Remove unused ([2ea8121](https://github.com/renatonerijr/switch-button-deploy/commit/2ea81210910b4f88426f6299743a8adf42d965d5))

# [1.1.0](https://github.com/renatonerijr/switch-button-deploy/compare/v1.0.0...v1.1.0) (2023-01-28)


### Features

* **export:** Export switch button ([4c5baff](https://github.com/renatonerijr/switch-button-deploy/commit/4c5baff51591addf89c15a9686dff5d07bb3e77c))

# 1.0.0 (2023-01-28)


### Performance Improvements

* **pencil:** add comments ([e03a7e7](https://github.com/renatonerijr/switch-button-deploy/commit/e03a7e7ca067306f045b9974c50bad7125dffe56))
* **pencil:** add package ([3f94d7b](https://github.com/renatonerijr/switch-button-deploy/commit/3f94d7bc8cd2308c53267714cb5ee8e15a7fb735))
* **pencil:** add package params ([1b52f25](https://github.com/renatonerijr/switch-button-deploy/commit/1b52f252e25f2c9a18503ae22967fed889844139))
* **pencil:** add package params ([5f8342d](https://github.com/renatonerijr/switch-button-deploy/commit/5f8342da43745e7fbd4fa36c0d2634e09a3e5cbd))
* **pencil:** add package params ([1c140bd](https://github.com/renatonerijr/switch-button-deploy/commit/1c140bd390f89b9282b4f2d321bd38714a80fe6d))
* **pencil:** add private false ([42b92a4](https://github.com/renatonerijr/switch-button-deploy/commit/42b92a4123e0be3b3fd93949f553d55eb30fd803))
* **pencil:** add private false ([30f661c](https://github.com/renatonerijr/switch-button-deploy/commit/30f661c0f070abaa94a5f54a7a1bcfb67e32d937))
* **pencil:** add private false ([a4c03a5](https://github.com/renatonerijr/switch-button-deploy/commit/a4c03a5713b048c5dae576d01f6c35c175df8c49))
* **pencil:** adjust words ([eb88ecb](https://github.com/renatonerijr/switch-button-deploy/commit/eb88ecbd15446f2b2638baff7957128af7f8f7b7))
* **pencil:** remove incorrect words ([5dbb650](https://github.com/renatonerijr/switch-button-deploy/commit/5dbb650f47b87ee5ae0d593bf63c536bf77119fa))

## [1.0.3](https://github.com/iwsouza/switch-button/compare/v1.0.2...v1.0.3) (2023-01-28)


### Performance Improvements

* **pencil:** add comments ([e03a7e7](https://github.com/iwsouza/switch-button/commit/e03a7e7ca067306f045b9974c50bad7125dffe56))
* **pencil:** add package params ([1b52f25](https://github.com/iwsouza/switch-button/commit/1b52f252e25f2c9a18503ae22967fed889844139))
* **pencil:** add package params ([5f8342d](https://github.com/iwsouza/switch-button/commit/5f8342da43745e7fbd4fa36c0d2634e09a3e5cbd))
* **pencil:** add package params ([1c140bd](https://github.com/iwsouza/switch-button/commit/1c140bd390f89b9282b4f2d321bd38714a80fe6d))
* **pencil:** add private false ([42b92a4](https://github.com/iwsouza/switch-button/commit/42b92a4123e0be3b3fd93949f553d55eb30fd803))
* **pencil:** add private false ([30f661c](https://github.com/iwsouza/switch-button/commit/30f661c0f070abaa94a5f54a7a1bcfb67e32d937))
* **pencil:** add private false ([a4c03a5](https://github.com/iwsouza/switch-button/commit/a4c03a5713b048c5dae576d01f6c35c175df8c49))

## [1.0.3](https://github.com/iwsouza/switch-button/compare/v1.0.2...v1.0.3) (2023-01-28)


### Performance Improvements

* **pencil:** add package params ([1b52f25](https://github.com/iwsouza/switch-button/commit/1b52f252e25f2c9a18503ae22967fed889844139))
* **pencil:** add package params ([5f8342d](https://github.com/iwsouza/switch-button/commit/5f8342da43745e7fbd4fa36c0d2634e09a3e5cbd))
* **pencil:** add package params ([1c140bd](https://github.com/iwsouza/switch-button/commit/1c140bd390f89b9282b4f2d321bd38714a80fe6d))
* **pencil:** add private false ([42b92a4](https://github.com/iwsouza/switch-button/commit/42b92a4123e0be3b3fd93949f553d55eb30fd803))
* **pencil:** add private false ([30f661c](https://github.com/iwsouza/switch-button/commit/30f661c0f070abaa94a5f54a7a1bcfb67e32d937))
* **pencil:** add private false ([a4c03a5](https://github.com/iwsouza/switch-button/commit/a4c03a5713b048c5dae576d01f6c35c175df8c49))

## [1.0.3](https://github.com/iwsouza/switch-button/compare/v1.0.2...v1.0.3) (2023-01-28)


### Performance Improvements

* **pencil:** add package params ([1b52f25](https://github.com/iwsouza/switch-button/commit/1b52f252e25f2c9a18503ae22967fed889844139))
* **pencil:** add package params ([5f8342d](https://github.com/iwsouza/switch-button/commit/5f8342da43745e7fbd4fa36c0d2634e09a3e5cbd))
* **pencil:** add package params ([1c140bd](https://github.com/iwsouza/switch-button/commit/1c140bd390f89b9282b4f2d321bd38714a80fe6d))
* **pencil:** add private false ([30f661c](https://github.com/iwsouza/switch-button/commit/30f661c0f070abaa94a5f54a7a1bcfb67e32d937))
* **pencil:** add private false ([a4c03a5](https://github.com/iwsouza/switch-button/commit/a4c03a5713b048c5dae576d01f6c35c175df8c49))


vite.config.ts
import react from '@vitejs/plugin-react';
import { defineConfig, loadEnv } from 'vite';
import { resolve } from 'path'

// https://vitejs.dev/config/
export default ({ mode }: { mode: string }) => {
  return defineConfig({
    plugins: [react()],
    define: {
      'process.env': { ...process.env, ...loadEnv(mode, process.cwd()) },
    },
    server: {
      port: 3000,
      watch: {
        usePolling: true,
      },
    },
    build: {
      lib: {
        // Could also be a dictionary or array of multiple entry points
        entry: resolve(__dirname, 'src/main.tsx'),
        name: 'SwitchButton',
        // the proper extensions will be added
        fileName: 'index',
      },
      rollupOptions: {
        // make sure to externalize deps that shouldn't be bundled
        // into your library
        external: ['react'],
        output: {
          // Provide global variables to use in the UMD build
          // for externalized deps
          globals: {
            react: 'React',
          },
        },
      },
    },
  });
};


src/components/switch-button/styles.ts
import styled, { css } from 'styled-components';

import { Props } from '.';

export const Wrapper = styled.button<{
  variant?: Props['variant'];
  isActive: boolean;
}>`
  position: relative;
  width: 24px;
  height: 50px;

  border-radius: 20px;
  overflow: hidden;

  border-radius: 20px;
  border: 1px solid #2f3740;
  background: #38424a;

  cursor: pointer;

  ${({ variant }) =>
    css`
      ${variant === 'horizontal' &&
      css`
        transform: rotate(-90deg);
      `}
      ${!variant &&
      css`
        transform: rotate(-90deg);
      `}
    `}

  &:disabled {
    opacity: 0.3;
    cursor: not-allowed;
  }
  &:before {
    content: '';
    position: absolute;
    width: 18px;
    height: 18px;
    background: #20272a;
    border: 1px solid #1e2426;
    border-radius: 50%;
    z-index: 2;
    left: 2px;
    top: 2px;
    transition: 0.3s cubic-bezier(0, 0.65, 0.36, 1.17) all;
  }
  ${({ isActive }) =>
    isActive &&
    css`
      &:before {
        top: 28.5px;
        background: #ffffff;
        border: 1px solid #ffffff;
      }
      background: #256a78;
      border: 1px solid #20606d;
    `}
`;


src/mapped.ts
export const steps = [
  {
    description: 'Instale as libs necessárias',
    codeString: `
  yarn add styled-components
  // ou
  npm install --save styled-components
  `,
  },

  {
    description: 'Crie o index.tsx e importe os estilos e coloque os tipos',
    codeString: `
    import * as S from './styles';

    export type Props = {
      variant?: 'horizontal' | 'vertical';
      isActive?: boolean;
      disabled?: boolean;
      onClick?: () => void;
    };
  `,
  },
  {
    description: 'Monte o componente SwitchButton',
    codeString: `
    /**
     * Componente Switch Button
     * @param variant horizontal | vertical
     * @param isActive boolean
     * @param disabled boolean
     * @param onClick return void function
     */
    export const SwitchButton: React.FC<Props> = ({
      variant,
      isActive,
      disabled,
      onClick,
    }: Props) => {
      return (
        <S.Wrapper
          variant={variant}
          isActive={!!isActive}
          disabled={disabled}
          onClick={onClick}
        />
      );
    };
  `,
  },
  {
    description: 'Crie o styles.ts e adicione os estilos',
    codeString: `
    import styled, { css } from 'styled-components';

    import { Props } from '.';

    export const Wrapper = styled.button<{
      variant?: Props['variant'];
      isActive: boolean;
    }>\`
        position: relative;
        width: 24px;
        height: 50px;

        border-radius: 20px;
        overflow: hidden;

        border-radius: 20px;
        border: 1px solid #2f3740;
        background: #38424a;

        cursor: pointer;

        \${({ variant }) =>
      css\`
          \${variant === 'horizontal' &&
          css\`
            transform: rotate(-90deg);
          \`}
          \${!variant &&
          css\`
            transform: rotate(-90deg);
          \`}
      \`}
      &:disabled {
        opacity: 0.3;
        cursor: not-allowed;
      }
      &:before {
        content: '';
        position: absolute;
        width: 18px;
        height: 18px;
        background: #20272a;
        border: 1px solid #1e2426;
        border-radius: 50%;
        z-index: 2;
        left: 2px;
        top: 2px;
        transition: 0.3s cubic-bezier(0, 0.65, 0.36, 1.17) all;
      }
      \${({ isActive }) =>
        isActive &&
          css\`
            &:before {
              top: 28.5px;
              background: #ffffff;
              border: 1px solid #ffffff;
            }
            background: #256a78;
            border: 1px solid #20606d;
          \`}
    \`;
    `,
  },
  {
    description: 'Agora é só criar um estado e importar o componente ',
    codeString: `
    export const SeuComponente: React.FC = () => {

      const [isActive, setIsActive] = useState(false);

      return (
        <SwitchButton
          variant="vertical"
          isActive={isActive}
          disabled
          onClick={() => setIsActive((old) => !old)}
        />
      );
    };
  `,
  },
];
